### ISSUE TITLE
Generators requiring temporaries live too long

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#44257]
This code:

```rust
fn foo() {
    use std::ops::{GeneratorState, Generator};

    || {
        let a = {
            let mut f = || {
                if false { yield }
                format!("a")
            };
            let a = loop {
                match f.resume() {
                    GeneratorState::Complete(e) => break e,
                    GeneratorState::Yielded(()) => {}
                }
                yield
            };
            a.len()
        };
        if false { yield }
        return a
    };
}
```

generates the error:

```
error[E0597]: `a` does not live long enough
   --> tests/smoke.rs:112:9
    |
111 |             a.len()
    |             - borrow occurs here
112 |         };
    |         ^ `a` dropped here while still borrowed
...
115 |     };
    |     - borrowed value needs to live until here

error: aborting due to previous error
```

but naively at least I'd expect it to work!

cc @Zoxc

**Comment by alexcrichton** [comment]
This may or may not be the same as issues like https://github.com/rust-lang/rust/issues/44197 or https://github.com/rust-lang/rust/issues/44200

**Comment by arielb1** [comment]
This looks like #44200 :

```Rust
fn foo() {
    use std::ops::{GeneratorState, Generator};

    || {
        let a = {
            let mut f = || {
                if false { yield }
                format!("a")
            };
            let a = loop {
                match f.resume() {
                    GeneratorState::Complete(e) => break e,
                    GeneratorState::Yielded(()) => {}
                }
                yield // this yield
            };
            a.len() // is caught up by this expression
        };
        if false { yield }
        return a
    };
}
```

**Comment by alex** [comment]
This was fixed by #44392
